### ISSUE TITLE
Grad strides do not match bucket view strides

### ISSUE THREAD

**Issue by ThibaultCastells** [pytorch/pytorch#83909]
### 🐛 Describe the bug

@mcarilli 
Hello,
I am using torch.nn.parallel.DistributedDataParallel and get the following warning:

> Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
> /\*\*\*\*\*\*\/miniconda3/envs/pym/lib/python3.10/site-packages/torch/autograd/__init__.py:173: UserWarning: Grad strides do not match bucket view strides. This may indicate grad was not created according to the gradient layout contract, or that the param's strides changed since DDP was constructed.  This is not an error, but may impair performance.
> grad.sizes() = [300, 100, 1, 1], strides() = [100, 1, 1, 1]
> bucket_view.sizes() = [300, 100, 1, 1], strides() = [100, 1, 100, 100] (Triggered internally at  /opt/conda/conda-bld/pytorch_1659484803030/work/torch/csrc/distributed/c10d/reducer.cpp:312.)

I could not write the minimal reproducing code because I can't find what part of the code create this warning. 
However, I know that it is due to an operation in my model as the warning does not appear with other models.
But I do not get any warning when running the same code with Apex, so could it be an issue with the native DDP?

### Versions

Collecting environment information...
PyTorch version: 1.12.1
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.3 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.27

Python version: 3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:35:26) [GCC 10.4.0] (64-bit runtime)
Python platform: Linux-5.4.0-120-generic-x86_64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090
GPU 2: NVIDIA GeForce RTX 3090
GPU 3: NVIDIA GeForce RTX 3090
GPU 4: NVIDIA GeForce RTX 3090
GPU 5: NVIDIA GeForce RTX 3090
GPU 6: NVIDIA GeForce RTX 3090
GPU 7: NVIDIA GeForce RTX 3090

Nvidia driver version: 515.48.07
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A  
Is XNNPACK available: True   

Versions of relevant libraries:
[pip3] numpy==1.23.2
[pip3] torch==1.12.1
[pip3] torchaudio==0.12.1
[pip3] torchvision==0.13.1   
[conda] blas                      2.116                       mkl    conda-forge
[conda] blas-devel                3.9.0            16_linux64_mkl    conda-forge
[conda] cudatoolkit               11.6.0              hecad31d_10    conda-forge
[conda] libblas                   3.9.0            16_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            16_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            16_linux64_mkl    conda-forge
[conda] liblapacke                3.9.0            16_linux64_mkl    conda-forge
[conda] mkl                       2022.1.0           h84fe81f_915    conda-forge
[conda] mkl-devel                 2022.1.0           ha770c72_916    conda-forge
[conda] mkl-include               2022.1.0           h84fe81f_915    conda-forge
[conda] numpy                     1.23.2          py310h53a5b5f_0    conda-forge
[conda] pytorch                   1.12.1          py3.10_cuda11.6_cudnn8.3.2_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                0.12.1              py310_cu116    pytorch
[conda] torchvision               0.13.1              py310_cu116    pytorch

**Comment by ngimel** [comment]
Minimum repro would help, but this warning is harmless. 
cc @alband - did you have a PR partially relaxing checks? Here it seems the warning is spurious too.

**Comment by ThibaultCastells** [comment]
Thanks! It's good to know that it's harmless, I was wondering if it could "impair performance" as the warning says.
I tried to identify the guilty operation, but as the issue occurs during back-propagation it is hard to tell.

**Comment by ThibaultCastells** [comment]
Thank you for the explanation. If this is the intended behavior, then explicitly silence the warning would be great indeed.

**Comment by bhack** [comment]
We have also still open:
https://github.com/pytorch/pytorch/issues/70222
https://github.com/pytorch/pytorch/issues/47163

**Comment by bhack** [comment]
I will add also https://github.com/pytorch/pytorch/issues/123560 to the list.

**Comment by bhack** [comment]
Can we check if https://github.com/pytorch/pytorch/issues/70222 https://github.com/pytorch/pytorch/issues/47163 and this one can be unified?

**Comment by shallweiwei** [comment]
Hello ,I have also faced this problem.
I made a cross attention layer like this, and I have used contiguous() after all of transpose or view code.
`class Cross_MultiAttention(nn.Module):
    def __init__(self,  emb_dim, num_heads):
        super(Cross_MultiAttention, self).__init__()
        self.emb_dim = emb_dim  # Dimension of the embedding is emb_dim
        self.num_heads = num_heads  # Number of attention heads is num_heads
        assert emb_dim % num_heads == 0, "emb_dim must be divisible by num_heads"
        self.depth = emb_dim // num_heads  # Dimension of each attention head is depth
        
        # Linear layers for query, key, value
        self.Wq = nn.Linear(emb_dim, emb_dim,bias=False)
        self.Wk = nn.Linear(768, emb_dim,bias=False)
        self.Wv = nn.Linear(768, emb_dim,bias=False)
        # Output projection layer
        self.proj_out = nn.Conv2d(emb_dim, emb_dim, kernel_size=1, stride=1, padding=0,bias=False)

    def forward(self, x, context):
        b, _, h, w = x.shape  # Input shape: [batch_size, emb_dim, height, width]
        # Reshape x to [batch_size, h*w, emb_dim]
        x = rearrange(x, 'b c h w -> b (h w) c').contiguous()
        # Compute query, key, value for self-attention computation
        Q = self.Wq(x).view(b, self.num_heads, -1,self.depth).contiguous( )  # [batch_size, num_heads, h*w, depth]
        K = self.Wk(context).view(b, self.num_heads,-1 ,self.depth).contiguous() # [batch_size, num_heads, seq_len, depth]
        V = self.Wv(context).view(b, self.num_heads,  -1,self.depth).contiguous()# [batch_size, num_heads, seq_len, depth]
        # Compute attention scores
        attn = (Q @ K.transpose(-2, -1)) / math.sqrt(self.depth)  # Output shape: [batch_size, num_heads, h*w, seq_len]
        attn = F.softmax(attn, dim=-1)  # Softmax over the last dimension to compute attention probabilities
        # Compute weighted sum of values
        out = attn @ V  # Output shape: [batch_size, num_heads, h*w, depth]
        out = rearrange(out, 'b head (h w) depth -> b (head depth) h w', h=h, w=w)
        # Project output to emb_dim
        out = self.proj_out(out)  # Output shape: [batch_size, emb_dim, height, width]

        return out`
I have used this layer 7 times in my model.
But I also faced another error. This just happen when I use this layer for the second time.
`Parameters which did not receive grad for rank 2: encoder_cross_layer2.norm3.body.bias, encoder_cross_layer2.norm3.body.weight, encoder_cross_layer2.ffn.project_out.weight, encoder_cross_layer2.ffn.dwconv.weight, encoder_cross_layer2.ffn.project_in.weight, encoder_cross_layer2.attn1.proj_out.weight, encoder_cross_layer2.attn1.Wv.weight, encoder_cross_layer2.attn1.Wk.weight, encoder_cross_layer2.attn1.Wq.weight, encoder_cross_layer2.norm1.body.bias, encoder_cross_layer2.norm1.body.weight`.
I'm really really confused about this

**Comment by shallweiwei** [comment]
I have confirmed this is not because of my implement of cross attention. I just changed this layer to a normal attention layer and I still get the same error

**Comment by bhack** [comment]
@albanD :point_up:

**Comment by albanD** [comment]
I'm afraid this is distributed-related questions so @wconstab would be the owner / best person to know who is the specific owner here.

**Comment by warptengood** [comment]
I know the warning is harmless, but it was a pain in my butt to see that thing on my logs. So this is what solved it in my case:

Environment: PyTorch 2.11, torch.compile + DistributedDataParallel

Symptom:
```
UserWarning: Grad strides do not match bucket view strides.
grad.sizes() = [1024, 576, 1, 1], strides() = [576, 1, 576, 576]
bucket_view.sizes() = [1024, 576, 1, 1], strides() = [576, 1, 1, 1]
```
Root cause:
torchinductor converts Conv2d operations to channels-last memory format internally when generating CUDA kernels. This means gradients for Conv2d weights arrive with channels-last strides (e.g. (576, 1, 576, 576)) during backward. DDP builds its gradient buckets by scanning parameter strides at construction time -- before any forward pass has run. If the model parameters are still in contiguous layout at that point, the bucket views are built with contiguous strides (576, 1, 1, 1), which don't match the channels-last grads that torchinductor later produces. DDP then falls back to an explicit copy on every backward step instead of a zero-copy view.

Workaround:
Convert model weights to channels-last before calling model.compile() and wrapping in DDP:
```
model = model.to(device=device, memory_format=torch.channels_last)
model.compile(mode='default')
model = DistributedDataParallel(model, device_ids=[...])
```
This ensures DDP builds its buckets with channels-last strides from the start, matching what torchinductor will produce during backward. As a secondary benefit, torchinductor no longer needs to insert a layout conversion op at the start of each forward pass.

Peace!
